pop-api/src/v0/assets/fungibles.rs

```diff
@@ -0,0 +1,422 @@
+///
+/// # Returns
+/// Returns `Ok(())` if successful, or an error if the operation fails.
+// fn destroy_approvals(id: AssetId) -> Result<()> {
+// 	Ok(dispatch(RuntimeCall::Assets(AssetsCall::DestroyApprovals {
+// 		id: id.into(),
+// 	}))?)
+// }
+
+/// Complete the process of destroying a token with a given asset ID.
+///
+/// # Arguments
+/// * `id` - The ID of the asset.
+///
+/// # Returns
+/// Returns `Ok(())` if successful, or an error if the operation fails.
+// fn finish_destroy(id: AssetId) -> Result<()> {
+// 	Ok(dispatch(RuntimeCall::Assets(AssetsCall::FinishDestroy {
+// 		id: id.into(),
+// 	}))?)
+// }
+
+/// Set the metadata for a token with a given asset ID.
+///
+/// # Arguments
+/// * `id` - The ID of the asset.
+///
+/// # Returns
+/// Returns `Ok(())` if successful, or an error if the operation fails.
+pub fn set_metadata(id: AssetId, name: Vec<u8>, symbol: Vec<u8>, decimals: u8) -> Result<()> {
+	assets::set_metadata(id, name, symbol, decimals)
+}
+
+/// Clear the metadata for a token with a given asset ID.
+///
+/// # Arguments
+/// * `id` - The ID of the asset.
+///
+/// # Returns
+/// Returns `Ok(())` if successful, or an error if the operation fails.
+// fn clear_metadata(id: AssetId) -> Result<()> {
+// 	Ok(dispatch(RuntimeCall::Assets(AssetsCall::ClearMetadata {
+// 		id: id.into(),
+// 	}))?)
+// }
+
+pub fn asset_exists(id: AssetId) -> Result<bool> {
+	assets::asset_exists(id)
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum FungiblesError {
+	Other(StatusCode),
+	/// The asset is not live; either frozen or being destroyed.
+	AssetNotLive,
+	/// Not enough allowance to fulfill a request is available.
+	InsufficientAllowance,
+	/// Not enough balance to fulfill a request is available.
+	InsufficientBalance,
+	/// The asset ID is already taken.
+	InUse,
+	/// Minimum balance should be non-zero.
+	MinBalanceZero,
+	/// The account to alter does not exist.
+	NoAccount,
+	/// The signing account has no permission to do the operation.
+	NoPermission,
+	/// The given asset ID is unknown.
+	Unknown,
+	// // TODO:
+	// // - Originally `InsufficientBalance` for the deposit but this would result in the same error
+	// // as the error when there is insufficient balance for transferring an asset.
+	/// No balance for creation of assets or fees.
+	NoBalance,
+}
+
+impl From<StatusCode> for FungiblesError {
+	fn from(value: StatusCode) -> Self {
+		let encoded = value.0.to_le_bytes();
+		match encoded {
+			// Balances.
+			[3, 10, 2, _] => FungiblesError::NoBalance,
+			// Assets.
+			[3, 52, 0, _] => FungiblesError::NoAccount,
+			[3, 52, 1, _] => FungiblesError::NoPermission,
+			[3, 52, 2, _] => FungiblesError::Unknown,
+			[3, 52, 3, _] => FungiblesError::InUse,
+			[3, 52, 5, _] => FungiblesError::MinBalanceZero,
+			[3, 52, 7, _] => FungiblesError::InsufficientAllowance,
+			[3, 52, 10, _] => FungiblesError::AssetNotLive,
```

Comment:
This seems quite hard to maintain. Ideally it would just be a from `StatusCode` or some overall flow like:
`StatusCode -> PopApiError -> FungiblesError`

Reply:
I disagree, a `From<StatusCode>` would mean I have to match on a `u32` which is visually harder to understand than the current implementation.

As for the `From<PopApiError>`, this feels less necessary to me. A user either wants to match on a `StatusCode` (default), a `PopApiError` (implemented with `From<StatusCode> for PopApiError`) or a `UseCaseError` (implemented with the above). When a `UseCaseError` is uncertain the user can convert the `Other(StatusCode)` into a `PopApiError`.

We could potentially offer a conversion method for the `Error::Module` into a `PalletError` or `UseCaseError`.

My suggestion is that I proceed with the other implementations for local fungibles, so that we have a better idea of what is needed, before I make changes to this implementation.

Reply:
Can the user define the error type on the extension? That way we can default it to StatusCode or pop_api::Error but they can override it based on their needs. That should help eliminate the .map_err().

We can provide impls for each error type for From StatusCode.

Reply:
E.g. https://github.com/r0gue-io/pop-node/blob/61b3a17d39b9bbbc76f87665a9546da9e0b58f89/pop-api/src/lib.rs#L49

I think you showed this on a call, but it would be nice if the contract developer could just specify the error type when using the extension, so they can opt in to choose their required level: 'use case' (easy), module, pop_api::error, StatusCode (advanced). Doing so could alleviate any .map_err(..) requirements depending on level.

Reply:
Definitely interesting but introduces extra features or code if no features are used. Suggest to look at this next sprint.

Reply:
matching it to a u32 doesn't decrease the binary size.